samples: more DT cleanups

mbolivar-nordic opened the pull request:
More incremental progress towards:

- dropping outmoded APIs like device_get_binding
- avoiding (driver-specific) DT_INST API usage in samples

As prep work:

Fixes: #31280

mbolivar-nordic committed:
samples: osdp/peripheral_device: drop device_get_binding

Use GPIO_DT_SPEC_GET_OR.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: task_wdt: drop device_get_binding/DT_INST

Use DEVICE_DT_GET and DT_COMPAT_GET_ANY_STATUS_OKAY.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: ipm_mhu_dual_core: drop device_get_binding/DT_INST

Use DEVICE_DT_GET_ANY.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: ipm_mcux: drop device_get_binding/DT_INST

Use DEVICE_DT_GET_ANY.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: ipm_mcux/remote: drop device_get_binding/DT_INST

Use DEVICE_DT_GET_ANY.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: ipm_imx: drop device_get_binding

Use DEVICE_DT_GET.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: osdp/control_panel: drop device_get_binding

Use GPIO_DT_SPEC_GET_OR.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: nvs: various cleanups

Replace device_get_binding with DEVICE_DT_GET(). Fix and clean up
error handling and debug printing during initialization.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
drivers: gpio: add GPIO_DT_SPEC_GET_OR and friends

These helper macros avoid boilerplate when constructing a gpio_dt_spec
structure from optional devicetree properties.

Update the release notes while we're here.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: hid-mouse: various cleanups

Use the new GPIO_DT_SPEC_GET_OR() macro to clean up the node detection
code by eliminating conditional compilation. This increases code
coverage on platforms where not all the devicetree nodes are
available.

Clean up some whitespace while we're at it.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: can: devicetree cleanups

Replace DT_CHOSEN_ZEPHYR_CAN_PRIMARY_LABEL with DEVICE_DT_GET on the
chosen node itself.

This is causing a build error in CI for the waveshare_open103z board
because it disables CAN by default in favor of USB. Throw in an
overlay reversing that decision for the sake of this sample to keep
things building. (Though the build previously succeeded, the
device_get_binding() call would simply have failed.)

Replace the LED access with GPIO_DT_SPEC_GET_OR. This changes the
semantics slightly so that the simulated LED implemented via printk()
is run if either the node is missing or the pin cannot be configured,
instead of just running when the node is missing. An unready LED port
when the node is defined still returns early from main.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: lora: drop device_get_binding

Use DEVICE_DT_GET.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: dts: bindings: add dt-zephyr-user target

Just a bit of future-proofing so we can continue to link to the right
place in case we ever want to move the /zephyr,user docs out of the
'inferred bindings' section in the DT guide.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: dac: move HW config to /zephyr,user

Require board overlays to configure the sample in the /zephyr,user
node, and access the device with DEVICE_DT_GET instead of
device_get_binding(). Update the documentation and move existing board
configuration to overlays.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: spi_fujitsu_fram: drop device_get_binding

Use DEVICE_DT_GET.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: entropy: drop device_get_binding

Use DEVICE_DT_GET.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: led_apa102: drop device_get_binding/DT_INST

Use DEVICE_DT_GET_ANY.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: peci: drop device_get_binding

Use DEVICE_DT_GET.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: led_ws2812: drop device_get_binding

Use DEVICE_DT_GET.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: led_lpd8806: drop device_get_binding/DT_INST

Use DEVICE_DT_GET_ANY.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: jesd216: drop device_get_binding/DT_INST

Use DEVICE_DT_GET/DT_COMPAT_GET_ANY_STATUS_OKAY.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: threads: devicetree cleanups

Use a gpio_dt_spec, GPIO_DT_SPEC_GET_OR, and DT_PROP_OR.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: button: devicetree cleanups

Use a gpio_dt_spec, dropping device_get_binding. This in turn lets us
write everything in a single main() without sacrificing readability
since the LED is optional.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: adxl372: drop device_get_binding/DT_INST

Use DEVICE_DT_GET_ANY.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: max30101: drop device_get_binding/DT_INST

Use DEVICE_DT_GET_ANY.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: mcp9808: drop device_get_binding/DT_INST

Use DEVICE_DT_GET_ANY.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: lis2dh: drop device_get_binding/DT_INST

Use DEVICE_DT_GET_ANY.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: fdc2x1x: drop device_get_binding/DT_INST

Use DEVICE_DT_GET/DT_COMPAT_GET_ANY_STATUS_OKAY.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: fxas21002: drop device_get_binding/DT_INST

Use DEVICE_DT_GET_ANY.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: tmp112: various cleanups

Add a real README, convert the device getter to DEVICE_DT_GET_ANY,
and enable assertions to make debugging easier on new users.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: bmg160: drop device_get_binding

Use DEVICE_DT_GET_ANY.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
samples: ms5837: drop device_get_binding/DT_INST

Use DEVICE_DT_GET_ANY.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

gmarull commented on include/drivers/gpio.h:
```diff
@@ -356,6 +356,25 @@ struct gpio_dt_spec {
+ * @return static initializer for a struct gpio_dt_spec for the property,
+ *         or default_value if the node or property do not exist
+ */
+#define GPIO_DT_SPEC_GET_OR(node_id, prop, default_value) \
```
I'm not sure if this macro will be ever used for cases other than

```c
GPIO_DT_SPEC_GET_OR(node_id, prop, {0})
```

So the macro could be designed without the need of explicitly defining it.

mbolivar-nordic replied:
> So the macro could be designed without the need of explicitly adding `{0}`.

Please review the discussion in  #31280; I'm doing it the way the user asked for it.

gmarull replied:
It's really a matter of taste, but I'd provide another macro that avoids passing `{0}` (which could be based on the current one).

mbolivar-nordic replied:
I'd rather not do it that way; I think the number of combinations will get too large when your PR #34539 is merged. This is very readable IMO and it is consistent with all the other DT-ish `THING_OR(node, prop, default)` APIs that explicitly take a `default` value.

gmarull commented on samples/subsys/usb/hid-mouse/src/main.c:
```diff
@@ -14,56 +14,42 @@
+ * Helper macro for initializing a gpio_dt_spec from the devicetree
+ * with fallback values when the nodes are missing.
+ */
+#define GPIO_SPEC(node_id) GPIO_DT_SPEC_GET_OR(node_id, gpios, {0})
```
see my previous comment

gmarull commented on samples/drivers/dac/src/main.c:
```diff
@@ -8,48 +8,32 @@
-#define DAC_DEVICE_NAME		DT_LABEL(DT_NODELABEL(dac0))
-#define DAC_CHANNEL_ID		0
-#define DAC_RESOLUTION		10
+#define ZEPHYR_USER_NODE DT_PATH(zephyr_user)
```
should we provide a pre-defined macro for such node, e.g. `DT_USER_NODE`?

mbolivar-nordic replied:
Maybe we can leave that to follow-up work? I'm not sure about that since we've always done it this way since inferred bindings for the `/zephyr,user` node were introduced, and I kind of like that this style makes it clear there's "no magic" from a devicetree.h API perspective.

gmarull replied:
sounds good

gmarull commented on samples/drivers/dac/src/main.c:
```diff
@@ -8,48 +8,32 @@
+#error "Unsupported board: see README and check /zephyr,user node"
+#define DAC_NODE DT_INVALID_NODE
+#define DAC_CHANNEL_ID 0
+#define DAC_RESOLUTION 0
```
are these definitions needed?

mbolivar-nordic replied:
Without them, you'll get build errors later down in the file where they are missing. With them, you just get the `#error` on line 20. This is more readable error output, which is important for a sample.

gmarull reviewed (commented):
Nice cleanup, I just have a few comments

de-nordic commented on samples/subsys/nvs/src/main.c:
```diff
@@ -68,25 +68,31 @@ void main(void)
 	 *	3 sectors
 	 *	starting at FLASH_AREA_OFFSET(storage)
 	 */
+	flash_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_flash_controller));
```
Shouldn't it be rather
`DEVICE_DT_GET(DT_MTD_FROM_FIXED_PARTITION(DT_NODE_BY_FIXED_PARTITION_LABEL(storage)))`
this way this would be really the device that holds the `storage` partition.
Eh we are missing define for that.

mbolivar-nordic replied:
>  Eh we are missing define for that.

Are you saying to leave this as-is? My goal with this patch is just to clean up the DT access, not to change semantics. This is currently using `device_get_binding(DT_CHOSEN_ZEPHYR_FLASH_CONTROLLER_LABEL)`, so this line is returning the exact same device.

de-nordic replied:
I think that this is bug, because the `zephyr_flash_controller` does not have to correspond to the flash controller the `storage` partition is under. So okay maybe replace it as you did and log an issue for investigation.

mbolivar-nordic replied:
OK, I can change it, let's see if @Laczen confirms

Laczen replied:
I can confirm this is a bug.

mbolivar-nordic replied:
Thanks @Laczen . I added a patch on top of the series that fixes the bug in the way @de-nordic suggested.

mbolivar-nordic committed:
samples: nvs: fix flash device access

Get the correct flash device from the storage partition node.
The current code is relying on the invalid assumption that the
chosen zephyr,flash-controller node is the device which contains
the storage partition.

Reported-by: Dominik Ermel <dominik.ermel@nordicsemi.no>
Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

henrikbrixandersen commented:
@mbolivar-nordic These are nice clean-ups, thanks. It's a bit much squashed into one PR - new macros, lots of samples touched, lots of commits. Why not split this PR into smaller PRs to ease the review process?

gmarull commented on include/drivers/gpio.h:
```diff
@@ -356,6 +356,25 @@ struct gpio_dt_spec {
+ * @return static initializer for a struct gpio_dt_spec for the property,
+ *         or default_value if the node or property do not exist
+ */
+#define GPIO_DT_SPEC_GET_OR(node_id, prop, default_value) \
```
Can you add `_OR` variants for the current set of initializers too?

mnkp commented on samples/drivers/lora/send/src/main.c:
```diff
@@ -25,13 +24,12 @@ char data[MAX_DATA_LEN] = {'h', 'e', 'l', 'l', 'o', 'w', 'o', 'r', 'l', 'd'};
-	if (!lora_dev) {
-		LOG_ERR("%s Device not found", DEFAULT_RADIO);
+	if (!device_is_ready(lora_dev)) {
+		LOG_ERR("%s Device not found", lora_dev->name);
```
We could change the message to "%s Device not ready", here and in a few more places in this PR.

mbolivar-nordic commented:
> @mbolivar-nordic These are nice clean-ups, thanks. It's a bit much squashed into one PR - new macros, lots of samples touched, lots of commits. Why not split this PR into smaller PRs to ease the review process?

@henrikbrixandersen This PR is doing one logical thing, namely cleaning up devicetree usage in samples. It's actually just a subset of the work that could be done in this area; we still have over 200 calls to device_get_binding() in the samples directory even with this PR. I really don't think it's worth splitting this up further; I'm just trying to get another chunk merged in time for feature freeze. I would have preferred to get all of the samples converted for 2.6 but there wasn't time.

I'm putting the new macros in as prep work commits so that they are not merged upstream without users.

mbolivar-nordic commented on samples/drivers/led_apa102/src/main.c:
```diff
@@ -56,11 +56,11 @@ void main(void)
-	if (strip) {
-		LOG_INF("Found LED strip device %s", DT_LABEL(DT_INST(0, apa_apa102)));
+	strip = DEVICE_DT_GET_ANY(apa_apa102);
+	if (strip && device_is_ready(strip)) {
```
> `if (device_is_ready(strip) {` ?

No, because DEVICE_DT_GET_ANY() may expand to NULL.

mbolivar-nordic commented on samples/drivers/can/src/main.c:
```diff
@@ -209,22 +199,20 @@ void main(void)
+			       led.port->name);
+			return;
+		}
+		ret = gpio_pin_configure_dt(&led, GPIO_OUTPUT_HIGH);
```
> I admit though it's a bit out of scope.

Yeah, let's take that to a follow-up PR.

mbolivar-nordic commented:
@gmarull @mnkp I've addressed your comments; thanks for review.

Other changes beyond those specific comments:

- get both lora samples (I accidentally only did lora/send; I meant to get lora/receive too)
- add release notes for all the new GPIO macros
- fix a copy/paste bug in the  threads sample (I had the wrong node when setting the gpio_pin_name for led1)

mbolivar-nordic commented:
Just a rebase to see if I can get green CI.

mbolivar-nordic commented:
Coding guidelines failure is spurious:

```
samples/subsys/nvs/src/main.c:51:WARNING: Violation to rule 5.1 or 5.2 (Identifiers shall be distinct) DT_NODE_BY_FIXED_PARTITION_LABEL length 32 > 31
```

mbolivar-nordic commented:
A couple of changes for CI; one is a real fix, one is a workaround to keep the build succeeding for `waveshare_open103z` in the can sample.

mbolivar-nordic commented:
CI failures seem spurious too:

```
/workdir/zephyr/drivers/sensor/max17262/max17262.c:343:13: error: Macro is deprecated [-Werror]
  343 | DT_INST_FOREACH_STATUS_OKAY(MAX17262_INIT)
```

I'm not touching this driver in this PR. It looks like a device_pm_control_nop issue.

mbolivar-nordic commented:
> It looks like a device_pm_control_nop issue.

Indeed it is, and was fixed in https://github.com/zephyrproject-rtos/zephyr/pull/34870
